src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,231 @@
+                dynamics[i][2] = Rx * signX[i / 2];
+            } else {
+                dynamics[i][0] = 0;
+                dynamics[i][1] = 1;
+                dynamics[i][2] = Ry * signY[i / 2];
+            }
+        }
+        Robot.gyro.reset();
+
+        if (!testMode) {
+            swerveModules[0] = new SwerveModule(0, DRIVE_MOTOR_0, ANGLE_MOTOR_0, FRONT_RIGHT_INVERTED);
+            swerveModules[1] = new SwerveModule(1, DRIVE_MOTOR_1, ANGLE_MOTOR_1, FRONT_LEFT_INVERTED);
+            swerveModules[2] = new SwerveModule(2, DRIVE_MOTOR_2, ANGLE_MOTOR_2, BACK_RIGHT_INVERTED);
+            swerveModules[3] = new SwerveModule(3, DRIVE_MOTOR_3, ANGLE_MOTOR_3, BACK_LEFT_INVERTED);
+        }
+
+        this.isFieldOriented = isFieldOriented;
+    }
+
+
+    /**
+     * sets the wheels of the robot to the calculated angle and speed
+     *
+     * @param forward  the Y value of the joystick
+     * @param strafe   the X value of the joystick
+     * @param rotation the rotation Z of the joystick
+     */
+    public void holonomicDrive(double forward, double strafe, double rotation) {
+
+        double[] robotHeading = getRobotHeading(strafe, forward, rotation, -Math.toRadians(Robot.gyro.getAngle()));
+
+        double[] velocities = calculateWheelVelocities(robotHeading);
+        double[] polar;
+        double[][] controls = new double[4][2];
+
+        // converts the cartesian velocities to polar and transfers them to a control matrix
+        for (int i = 0; i < 4; i++) {
+            polar = Utils.cartesianToPolar(velocities[2 * i + 1], velocities[2 * i]);
+            controls[i][0] = polar[0];
+            controls[i][1] = polar[1];
+        }
+
+        // feeds the corresponding control to each wheel
+        for (int k = 0; k < 4; k++) {
+            swerveModules[k].setSpeed(controls[k][0]);
+            swerveModules[k].setAngle(controls[k][1]);
+        }
+
+        double sumx = 0;
+        double sumy = 0;
+        for (int j = 0; j < 4; j++) {
+            sumx += velocities[j * 2];
+            sumy += velocities[j * 2 + 1];
+        }
+        double[] target = Utils.cartesianToPolar(sumx, sumy);
+        FireLog.log("target velocity", target[0]);
+        FireLog.log("target angle", target[1]);
+    }
+
+    /**
+     * turns the joystick inputs into the robot heading
+     *
+     * @param forward    the Y value of the joystick
+     * @param strafe     the X value of the joystick
+     * @param rotation   the rotation Z of the joystick
+     * @param robotAngle the current angle of the robot in radians
+     * @return an array of the robot heading
+     */
+    public static double[] getRobotHeading(double forward, double strafe, double rotation, double robotAngle) {
+        // turns the joystick values into the heading of the robot
+        forward *= Constants.SwerveDrive.SPEED_MULTIPLIER;
+        strafe *= Constants.SwerveDrive.SPEED_MULTIPLIER;
+        rotation *= Constants.SwerveDrive.ROTATION_MULTIPLIER;
```

Comment:
IMO, these multiplications should happen in the command, and not here. This method shouldn't get "joystick values", but commands for each axis. I believe you want to use this method in autonomous driving, no? If yes, you'd probably not want these multiplications (which will be a bug and not just a design issue).